scripts/word_embeddings/train_fasttext.py

```diff
@@ -260,106 +290,155 @@ def train(args):
+                    else:
+                        emb_in = embedding(center)
+
+                    with mx.autograd.pause():
+                        word_context_negatives = mx.nd.concat(
+                            word_context, negatives, dim=1)
+                        word_context_negatives_mask = mx.nd.concat(
+                            word_context_mask, negatives_mask, dim=1)
 
                     emb_out = embedding_out(word_context_negatives,
                                             word_context_negatives_mask)
 
                     # Compute loss
-                    pred = mx.nd.batch_dot(
-                        emb_in.expand_dims(1), emb_out.swapaxes(1, 2))
+                    pred = mx.nd.batch_dot(emb_in, emb_out.swapaxes(1, 2))
                     pred = pred.squeeze() * word_context_negatives_mask
                     label = mx.nd.concat(word_context_mask,
                                          mx.nd.zeros_like(negatives), dim=1)
 
                 elif args.model.lower() == 'cbow':
-                    emb_in = embedding(word_context, word_context_mask,
-                                       subwords, subwords_mask).sum(axis=-2)
-
-                    center_negatives = mx.nd.concat(
-                        center.expand_dims(1), negatives, dim=1)
-                    center_negatives_mask = mx.nd.concat(
-                        center_mask.expand_dims(1), mx.nd.ones_like(negatives),
-                        dim=1)
+                    word_context = word_context.reshape((-3, 1))
+                    word_context_mask = word_context_mask.reshape((-3, 1))
+                    if args.ngram_buckets:
+                        emb_in = embedding(word_context, subwords,
+                                           word_context_mask, subwords_mask,
+                                           inverse_unique_indices)
+                    else:
+                        emb_in = embedding(word_context, word_context_mask)
+
+                    with mx.autograd.pause():
+                        center = center.tile(args.window * 2).reshape((-1, 1))
+                        negatives = negatives.reshape((-1, args.negative))
+
+                        center_negatives = mx.nd.concat(
+                            center, negatives, dim=1)
+                        center_negatives_mask = mx.nd.concat(
+                            mx.nd.ones_like(center), negatives_mask, dim=1)
 
                     emb_out = embedding_out(center_negatives,
                                             center_negatives_mask)
 
                     # Compute loss
-                    pred = mx.nd.batch_dot(
-                        emb_in.expand_dims(1), emb_out.swapaxes(1, 2))
-                    pred = pred.reshape((-1, 1 + args.negative))
+                    pred = mx.nd.batch_dot(emb_in, emb_out.swapaxes(1, 2))
+                    pred = pred.squeeze() * word_context_mask
```

Comment:
I understand we want to minimize the dot(l(s(w_t,w_c)) + l(-s(w_t,w_c))), but what do we need word_context_mask mean here?